doc/rtd/howto/pgp.rst

```diff
@@ -0,0 +1,200 @@
+
+Our encrypted and signed user data is now saved in `cloud-config.yaml.asc`.
+
+Export the keys
+===============
+
+In order to use this user data, we will need to create a custom image
+containing the public key of the encrypting user and the private key
+of the signing user.
+
+Create the key directory:
+
+.. code-block:: bash
+
+    $ mkdir /etc/cloud/keys
+
+Export the public key of the signing user:
+
+.. code-block:: bash
+
+    $ gpg --export signing_user > /etc/cloud/keys/signing_user.gpg
+
+Export the private key of the encrypting user:
+
+.. code-block:: bash
+
+    $ gpg --export-secret-keys encrypting_user > /etc/cloud/keys/encrypting_user.gpg
+
+Why export keys?
+----------------
+
+While it is more steps to export the keys in this way as opposed to
+using the existing key ring in the snapshot, we do this for a few reasons:
+
+* Users may not want these keys in any key ring by default on a new instance
+* Exporting keys is easier than copying key rings
+
+Note that on launch, cloud-init will import these keys into a temporary
+key ring that is removed after the user data is processed. The default
+key ring will not be read or modified.
+
+Restrict user data to require PGP message
+=========================================
+
+To ensure that our message hasn't been replaced or tampered with, we can
+require that cloud-init only process PGP messages. To do so, create a file
+`/etc/cloud/cloud.cfg.d/99_pgp.cfg` with the following contents:
+
+.. code-block:: text
+
+    user_data:
+      require_signature: true
+
+Retrieve our encrypted and signed user data
+===========================================
+
+Before running
+these commands, copy the encrypted and signed user data
+that we created earlier to the host system.
+
+From the host system, run:
+
+.. code-block:: bash
+
+    $ lxc file pull pgp-demo/root/cloud-config.yaml.asc .
+
+
+Create a custom image containing the keys
+=========================================
+
+.. note::
+    Before creating the image, you may want to remove the original user data
+    and created key ring from the instance. This is not strictly necessary
+    but is recommended for a clean image.
+
+Now that we have our instance configured, we can create a custom image from
+it. This step will vary depending on your cloud provider.
+
+Using LXD, from the host system, run:
+
+.. code-block:: bash
+
+    $ lxc stop pgp-demo
+    $ lxc publish pgp-demo --alias pgp-demo-image
+
+Launch an instance with the encrypted user data
+===============================================
+
+Now that we have our custom image with the keys, we can launch a
+new instance with the encrypted user data. With your encrypted and signed
+user data in the current working directory, run:
+
+.. code-block:: bash
+
+    $ lxc launch pgp-demo-image pgp-demo-encrypted \
+      --config user.user-data="$(cat cloud-config.yaml.asc)"
+
+On the launched system, you should see the file `/var/tmp/hello-world.txt`
```

Comment:
I just ran through this tutorial (making sure to remove the gpg keys from the root keychain), but it failed for me:

```python
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/gpg.py", line 106, in decrypt
    subp.subp(
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 291, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['gpg', '--verify']
Exit code: 2
Reason: -
Stdout: 
Stderr: gpg: verify signatures failed: Unexpected error

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 820, in status_wrapper
    ret = functor(name, args)
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 487, in main_init
    init.update()
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 570, in update
    self._store_processeddata(self.datasource.get_userdata(), "userdata")
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 650, in get_userdata
    self.userdata = self.ud_proc.process(
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/user_data.py", line 98, in process
    convert_string(
  File "/usr/lib/python3/dist-packages/cloudinit/user_data.py", line 452, in convert_string
    bdata = handle_encrypted(bdata, is_part, require_signature)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/user_data.py", line 413, in handle_encrypted
    return decrypt_payload(data.decode("utf-8"), require_signature).encode(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/user_data.py", line 393, in decrypt_payload
    return gpg_context.decrypt(
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/gpg.py", line 113, in decrypt
    raise GpgVerificationError(
cloudinit.gpg.GpgVerificationError: Signature verification failed
```